src/server/storage_engine/block_engine/allocator.rs

```diff
@@ -2,54 +2,317 @@
+                            if borrow_size == 0 {
+                                borrow_size = chunk_size;
+                                chunk_size -= 1;
+                            } else {
+                                chunk_size = borrow_size - 1;
+                                if_need_borrow_from_smaller = true;
+                            }
+                        } else {
+                            let mut iter = slice_space_vec[chunk_size as usize].iter();
+                            let mut begin = 0;
+                            if let Some(slice) = iter.next() {
+                                begin = *slice;
+                            }
+                            if borrow_size != 0 {
+                                if size >= borrow_size {
+                                    self.release_space(borrow_size, begin);
+                                    self.release_space(chunk_size - borrow_size, begin);
+                                    let alloc = AllocatorEntry {
+                                        begin,
+                                        length: borrow_size,
+                                    };
+                                    result_vec.push(alloc);
+                                }
+                            } else {
+                                let alloc = AllocatorEntry {
+                                    begin,
+                                    length: chunk_size,
+                                };
+                                result_vec.push(alloc);
+                                slice_space_vec[chunk_size as usize].remove(&begin);
+                            }
+                            break;
+                        }
+                    }
+                    _ => {
+                        let mut size = chunk_size;
+                        let mut allocator_size = 10;
+                        loop {
+                            if size - allocator_size <= 10 {
+                                break;
+                            }
+                            if slice_space_vec[allocator_size as usize].is_empty() {
+                                allocator_size -= 1;
+                            } else {
+                                let mut iter = slice_space_vec[allocator_size as usize].iter();
+                                let mut begin = 0;
+                                if let Some(slice) = iter.next() {
+                                    begin = *slice;
+                                }
+                                let alloc = AllocatorEntry {
+                                    begin,
+                                    length: allocator_size,
+                                };
+                                result_vec.push(alloc);
+                                slice_space_vec[allocator_size as usize].remove(&begin);
+                                size -= allocator_size;
```

Comment:
Why not set size as (size - 2^allocator_size)? Do you mean the allocator_size is real size rather than scale?